<a href="https://colab.research.google.com/github/GayanaJayasinghe/tutorials/blob/master/Lesson1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Baby steps


Let's try printing Hello world.

In [ ]:
print("Hello world!")

Hello world!


Congratulations! This is your first code. Now lets try changing it a little.

In [ ]:
print(Hello world!)

SyntaxError: invalid syntax. Perhaps you forgot a comma? (<ipython-input-2-01a3c6a17cb8>, line 1)

This didn't work. What you have above is called an error message. That is because the computer cannot **interpret** the **instruction** that was just used. When we enter a command to print with values, we need to insert quotes. This denotes that the type of information we are entering is a string. Let us try a sequence of commands to test this.


In [ ]:
A="Hello world!"
print(A)

Hello world!


What happened here? We first declare a variable A, and set it's value to be a string "Hello world!". We can sheck this as follows.

In [ ]:
A="Hello world!"
print(type(A))
print(A)

<class 'str'>
Hello world!


The instruction type(A) takes the type of the variable A and it is printed by the print command.
Lets try and see the type of another variable.

In [ ]:
B=23
print(type(B))
print(B)

<class 'int'>
23


The first line of the three above declares a variable B and stores the value 23 in it. It is stored as an integer type variable, which is shortened and denoted as 'int'.
The second line prints this.
The third line prints the value of B.

Lets try another version of this.

In [ ]:
A="31"
B=23
print(type(A))
print(type(B))
print(A)
print(B)

<class 'str'>
<class 'int'>
31
23


Can you interpret what happens in these lines?
Look at the earlier examples and compare.

Let us now try another set of instructions.

In [ ]:
A="31"
B=23
C=45
D="55"
print(B+C)
print(A+D)
print(A+B)

68
3155


TypeError: can only concatenate str (not "int") to str

What happened here? First the code returned 68, which is the sum of the integers B and C. Then in the second line it returned what we call the concatenation of the two strings A and D. Basically it combined the two strings "31"+"55" and gave the new string variable "3155". Then it tried to do either summing or concatenating A and B, but since they are of different data types, there is no operation in python that adds such data types together, so it returned an error.

Lets explore this with the following code.

In [ ]:
A="31"
B=23
C=45
D="55"
print(B+C)
print(type(B+C))
print(A+D)
print(type(A+D))

68
<class 'int'>
3155
<class 'str'>


We can also do the following variations.

In [ ]:
A="31"
B=23
C=45
D="55"
print(str(B)+str(C))
print(int(A)+int(D))

2345
86


So the instruction + is said to be overloaded. It means that is means different things based on context. Even in human conversations and human languages used for communication, many things depend on context. It is the same with programming languages.

A computer can check if certain statements are true or false. It can compare if two quantities are equal or not.

In [ ]:
A=2
B=3
C=3
#The next line checks if the values in A and B are equal, and stores it in a boolean variable internally, the value of which is printed.
print(A==B)
#The next line checks if the values in A and B are NOT equal, and stores it in a boolean variable internally, the value of which is printed.
print(A!=B)
#The next line checks if the values in B and C are equal, and stores it in a boolean variable internally, the value of which is printed.
print(B==C)
#The next line checks if the values in B and C are NOT equal, and stores it in a boolean variable internally, the value of which is printed.
print(B!=C)

#The next line checks if the values in A and B are equal, and stores it in a boolean variable D.
D=(A==B)
#The next line prints the type of the variable D
print(type(D))
#The next line prints the value of D
print(D)

False
True
True
False
<class 'bool'>
False


We have added comments to the code which explains each line of code. Try adjusting the values of A, B, C and see what happens.

The computer can also do things based on whether not the value of a boolean statement is true or false. Consider the following code snippets.

In [ ]:
if True:
  print(1)
else:
  print(2)

1


In [ ]:
if False:
  print(1)
else:
  print(2)

2


In [ ]:
A=2
B=3
if A+B==5:
  print(1)
else:
  print(2)


1


In the last code snippet, two variables A and B are declared and the values A=2 and B=3 are stored within it. Then in the if condition, it is checked if A+B==5, which returns a boolean value, True or False. If True is returned, then the instructions below that if condition are executed. Here is another variation of the above code snippet.

In [ ]:
A=2
B=3
C=A+B
if C==5:
  print("Bright skies")
else:
  print("Darkness")


Bright skies


Here's another.

In [ ]:
A="App"
B="le"
C=A+B
if C=="Apple":
  print("Bright skies")
else:
  print("Darkness")


Bright skies


And another

In [ ]:
A="App"
B="le"
C=A+B
if C=="Appe":
  print("Bright skies")
else:
  print("Darkness")

Darkness


We observe that the number of spaces, or tabs is important. This is called indentation, and the python language needs proper indentation. Let's try a code without proper indentation.

In [ ]:
A="App"
B="le"
C=A+B
if C=="Appe":
print("Bright skies")
else:
  print("Darkness")

IndentationError: expected an indented block after 'if' statement on line 4 (<ipython-input-9-ff8ed50245b2>, line 5)

So we have to be careful with the data types and indenation, so that the computer understands what we are telling it. Let's see if we can put these together to have a useful code.

In [ ]:
Cricket_team_1_total=245
Cricket_team_2_total=250
if Cricket_team_1_total>Cricket_team_2_total:
  print("Team 1 wins")
else:
  print("Team 2 wins")

Team 2 wins


How do you handle the case if there is a draw? There are two ways to do it. One is to use another if else statement.

In [ ]:
Cricket_team_1_total=245
Cricket_team_2_total=245
if Cricket_team_1_total>Cricket_team_2_total:
  print("Team 1 wins")
else:
  if Cricket_team_1_total==Cricket_team_2_total:
    print("Draw")
  else:
    print("Team 2 wins")

Draw


What we have above is an if condition within another else condition. We call these nested conditionals. For the above problem we have simply used it to check multiple conditions. However that can be handled using the elif condition (shortened for the logical statement else if).

In [ ]:
Cricket_team_1_total=245
Cricket_team_2_total=245
if Cricket_team_1_total>Cricket_team_2_total:
  print("Team 1 wins")
elif Cricket_team_1_total==Cricket_team_2_total:
  print("Draw")
else:
  print("Team 2 wins")

Draw


Let's look at the following code, which we can use to take user inputs, and then print it.


In [ ]:
user_input = input("User input: ")
print(user_input)

User input: Print me
Print me


We can use this command to build a small code to check who won, given the scores of two teams. Let's now ask the AI to generate this for us.

In [ ]:
# prompt: Take two user inputs called "Total of team 1" and "Total of team 2"

team1_total = input("Total of team 1: ")
team2_total = input("Total of team 2: ")


The AI can be thought of as a helper. You the human wants to write a code. But it takes time to type code. And you have to be very specific, and use the correct syntax when coding. The AI is like a very smart coder, who can interpret very high level prompts, similar to humans. When we tell a human to "write a code that takes two variables called a, and b, and return the difference of them", a human will take some time and write that code. The AI helper does this in a shorter time. Later in the course we will discuss how this is achieved if we have time. For now, lets try the following prompt.

In [ ]:
# prompt: write a code that takes two variables called a, and b, and return the difference of them

def difference(a, b):
  return a - b


So the AI generated a code which is a bit more advanced than what we were looking for. Sometimes even humans do not understand us the first time, and we have to describe things in more detail. Let's ask it again to try and get a simpler code, by describing what we want in more detail.

In [ ]:
# prompt: Take two user inputs, called a and b. Then take their difference, store it in a variable c. Then print c

a = int(input("Enter a: "))
b = int(input("Enter b: "))
c = a - b
print(c)


That's much better. You can ask the AI to write code for you. But make sure you understand what it is doing at each step.

Now let us try modifying the code we had earlier.

In [ ]:
team1_total = input("Total of team 1: ")
team2_total = input("Total of team 2: ")
if team1_total>team2_total:
  print("Team 1 wins")
elif team1_total==team2_total:
  print("Draw")
else:
  print("Team 2 wins")


Total of team 1: 45
Total of team 2: 45
Draw


Lets try prompting the AI for this code, or something similar. We will write in "natural language", or human friendly language, and the AI will translate it to a computer friendly language, or to a code in the python language.

In [ ]:
# prompt: Take two user inputs called team1_total and tean2_total corresponding to the teams 1 and 2. Write a code that prints which team has won, or whether the match is drawn

team1_total = int(input("Total of team 1: "))
team2_total = int(input("Total of team 2: "))
if team1_total > team2_total:
  print("Team 1 wins")
elif team1_total == team2_total:
  print("Draw")
else:
  print("Team 2 wins")


You see that the AI gave us the same code as above. This is no coincidence. The AI remembers everything in this sheet, and uses the context coming from previous prompts to understand how the human has been prompting. Notice that in my prompt we added a spelling mistake "tean2_total" rather than "team2_total". This was a wilful error that I made. Auto correcting aside, it reads what we said carefully, and even reads the text that is written here.
Let's try a different code.

In [ ]:
# prompt: Take three inputs called length, width and height. Compute the volume and store it in a variable. Then print it.

length = int(input("Enter length: "))
width = int(input("Enter width: "))
height = int(input("Enter height: "))
volume = length * width * height
print(volume)


Enter length: 3
Enter width: 4
Enter height: 3
36


We did not even have to say how to compute the volume. It found an example from some data on the internet (or when it was trained), and even says where it got it from. Notice how it came up with suitable variable names which makes things much clearer to read. This is a good coding practice.
Now lets try another code.

In [ ]:
# prompt: Write a code that can compute the volume of a cylinder and print it, after taking inputs radius and height

import math

radius = float(input("Enter radius: "))
height = float(input("Enter height: "))
volume = math.pi * radius**2 * height
print(volume)


Enter radius: 3
Enter height: 4
113.09733552923255


Can you understand the above code? What is float? Basically its a data type that allows decimal places.
We will discuss that data type later in more detail. The symbol * is multiplication. The symbol ** is the power. So radius**2 is the square of the radius.

It also imported math.pi. We will discuss what this means later. Let's try to ask the AI to write a code without importing math.

In [ ]:
# prompt: Write a code that can compute the volume of a cylinder and print it, after taking inputs radius and height. Don't import anything.

radius = float(input("Enter radius: "))
height = float(input("Enter height: "))
pi = 3.14159265359
volume = pi * radius**2 * height
print(volume)


So now it approximated pi.

Let's try and generate one more code with if else conditions.

In [ ]:
# prompt: Take 3 user inputs called height of John, height of Joan, height of William. If william is taller than Joan, print it. If Joan is taller than william print it

height_john = int(input("Enter height of John: "))
height_joan = int(input("Enter height of Joan: "))
height_william = int(input("Enter height of William: "))
if height_william > height_joan:
  print("William is taller than Joan")
elif height_joan > height_william:
  print("Joan is taller than William")


Note that it is taking the heights to be integers. We want to measure height in meters, but with decimal places. So we need to use float. We can modify this ourselves since we understand it.

In [ ]:
# prompt: Consider the heights of the people obtained with user inputs, sort them in descending order using if else condition, the print their names in descending order

height_john = float(input("Enter height of John: "))
height_joan = float(input("Enter height of Joan: "))
height_william = float(input("Enter height of William: "))

if height_john >= height_joan and height_john >= height_william:
  tallest = "John"
  if height_joan >= height_william:
    second_tallest = "Joan"
    shortest = "William"
  else:
    second_tallest = "William"
    shortest = "Joan"
elif height_joan >= height_john and height_joan >= height_william:
  tallest = "Joan"
  if height_john >= height_william:
    second_tallest = "John"
    shortest = "William"
  else:
    second_tallest = "William"
    shortest = "John"
else:
  tallest = "William"
  if height_john >= height_joan:
    second_tallest = "John"
    shortest = "Joan"
  else:
    second_tallest = "Joan"
    shortest = "John"

print(tallest)
print(second_tallest)
print(shortest)


Enter height of John: 1.6
Enter height of Joan: 1.5
Enter height of William: 1.3
John
Joan
William


See how easy that was? I didn't even have to say the names again because the AI can see the names i used in the previous prompt. It is also reading even these lines in the text. See if you can understand every step of the code.
The "and" boolean operator checks if two boolean variables are both true, and returns a boolean value true, and if not it returns false. We can check it like this.

In [ ]:
print(True and True)
print(True and False)
print(False and True)
print(False and False)


True
False
False
False


We can now revert to the lecture notes where we will review some of the things we have studied using pseudocode.

# UNDER CONSTRUCTION.

DO WE FIRST DO LOOPS? Or we can introduce for loops only in order to iterate through a loop. Arrays, lists, and dictionaries

In this section we will study how we can store various sets of data that are related to each other, rather than storing various quantities in individual variables. We shall see how this allows us to think of various associated operations in simpler terms, and use built in functions in the python language and its libraries to handle large data sets more easily. We begin by studying arrays.


In [ ]:
# prompt: Create a new array to store the height of 5 boxes. Take a user input for the boxes as an array. Sort the heights of the boxes.

height_of_boxes = []
for i in range(5):
  height = float(input("Enter height of box " + str(i+1) + ": "))
  height_of_boxes.append(height)
height_of_boxes.sort()
print(height_of_boxes)


Enter height of box 1: 22
Enter height of box 2: 33
Enter height of box 3: 32
Enter height of box 4: 52
Enter height of box 5: 11
[11.0, 22.0, 32.0, 33.0, 52.0]


Let us consider the height based name sorting problem that we considered in a previous lesson. We have three people with three heights, and we want to sort the names according to height.
One way to store the names and the heights is in an array as given below.

In [ ]:
Names=["John","Joan","William"]
Heights=[1.5, 1.4, 1.6]
print(Names)
print(Heights)

['John', 'Joan', 'William']
[1.5, 1.4, 1.6]


We can get only the name and height of one person.

In [ ]:

print(Names[0])
print(Heights[0])

John
1.5


In [ ]:
print(Names[1])
print(Heights[1])

Joan
1.4


Notice that the 1st set of entries are labelled, or indexed, by 0. The second set of entries are labelled by 1. The counter of for indexing arrays and other data structures starts by 0 in python.

We can add a new name and a new height to each of these using the append command.

In [ ]:
Names.append("James")
Heights.append(1.7)
print(Names)
print(Heights)

['John', 'Joan', 'William', 'James']
[1.5, 1.4, 1.6, 1.7]


In [ ]:
Names.append("James")
Heights.append(1.4)
print(Names)
print(Heights)

['John', 'Joan', 'William', 'James', 'James']
[1.5, 1.4, 1.6, 1.7, 1.4]


Observe that we can have multiple entries. In fact the two arrays are not directly related in the machine and we can treat them separately.

In [ ]:
Names.append("Jane")
print(Names)
print(Heights)

['John', 'Joan', 'William', 'James', 'James', 'Jane']
[1.5, 1.4, 1.6, 1.7, 1.4]


Lets take a subset of the above array. Let's take the 1st three entries and make a new array called Names_1

In [ ]:
Names_1=Names[0:3]
print(Names_1)
print(Names[3])

['John', 'Joan', 'William']
James


We used 0:3, so it "sliced" the array and took the entries starting from the index 0, upto the index 3, but without the 4th index which is labelled by 3, which was "William".

Question:Can you try slicing a different subset?
Try a subset with the 2nd and the third entry.

In [ ]:
Names=["John","Joan","William"]
Heights=[1.5, 1.4, 1.6]
Names_2=Names[1:3]
print(Names_2)

['John', 'Joan', 'William']
[1.5, 1.4, 1.6]
['Joan', 'William']


How about taking two subsets and combing them.
Let's use a simpler example for manipulations. I will declare an array A with 6 numbers and take two subsets, of the first two and the last two numbers. Then I will combine them.

In [ ]:
A=[1,2,3,4,5,6]
A_1=A[0:2]
A_2=A[4:6]
A_3=A_1+A_2
print(A_3)


[1, 2, 5, 6]


So this is another operation for which the symbol is +.

In [ ]:
A=["s","t"]
B=["a","b"]
C=A+B
print(C)

['s', 't', 'a', 'b']


Arrays are very flexible. They can store different data types within even the same array.

In [ ]:
A=["s",1]
print(A[0])
print(A[1])

s
1


I feel like this has become too technical somehow. Maybe its best to introduce nicer, more useful code, before explaining more abstract code.
Not everyone will appreciate the power of abstraction initially.

# Construction. Loops

In many day to day activities, we have to do somethings repeatedly. For instance if we are taking the marks of students in a class and finding the class average various subjects, we have to take the average of values for each subject.
This is one instance where loops help us. Let's begin with a simple example of finding class averages.

In [ ]:
# prompt: Create a dictionary with three students and their marks for a subject maths

student_marks = {
    "Alice": 85,
    "Bob": 72,
    "Charlie": 90
}
print(student_marks)


In [ ]:
# prompt: Now compute the average marks for maths

student_marks = {
    "Alice": 85,
    "Bob": 72,
    "Charlie": 90
}
total_marks = 0
for student in student_marks:
  total_marks += student_marks[student]
average_marks = total_marks / len(student_marks)
print("Average marks for maths:", average_marks)
